src/include/cursor.i

```diff
@@ -321,6 +336,8 @@ __cursor_func_init(WT_CURSOR_BTREE *cbt, bool reenter)
 
 	session = (WT_SESSION_IMPL *)cbt->iface.session;
 
+	F_CLR(cbt, WT_CBT_RETRY_NEXT | WT_CBT_RETRY_PREV);
+
```

Comment:
I think a simpler/cleaner approach to this would be to clear `WT_CBT_RETRY_NEXT` in `__wt_btcur_next` and `WT_CBT_RETRY_PREV` in `__wt_btcur_prev`. Then `__cursor_func_init()` doesn't do anything at all with those flags, (either testing or clearing). Something like:
```
diff --git a/src/btree/bt_curnext.c b/src/btree/bt_curnext.c
index 82096bb3c..25986779f 100644
--- a/src/btree/bt_curnext.c
+++ b/src/btree/bt_curnext.c
@@ -610,6 +610,7 @@ __wt_btcur_next(WT_CURSOR_BTREE *cbt, bool truncating)
                        F_CLR(cbt, WT_CBT_RETRY_NEXT);
                        return (__cursor_kv_return(session, cbt, upd));
                }
+               F_CLR(cbt, WT_CBT_RETRY_NEXT);
        }
 
        WT_RET(__cursor_func_init(cbt, false));
diff --git a/src/btree/bt_curprev.c b/src/btree/bt_curprev.c
index 7e5783a8e..aa2b8918b 100644
--- a/src/btree/bt_curprev.c
+++ b/src/btree/bt_curprev.c
@@ -565,6 +565,7 @@ __wt_btcur_prev(WT_CURSOR_BTREE *cbt, bool truncating)
                        F_CLR(cbt, WT_CBT_RETRY_PREV);
                        return (__cursor_kv_return(session, cbt, upd));
                }
+               F_CLR(cbt, WT_CBT_RETRY_PREV);
        }
 
        WT_RET(__cursor_func_init(cbt, false));
diff --git a/src/include/cursor.i b/src/include/cursor.i
index 4e4430e4c..a8d739514 100644
--- a/src/include/cursor.i
+++ b/src/include/cursor.i
@@ -336,8 +336,6 @@ __cursor_func_init(WT_CURSOR_BTREE *cbt, bool reenter)
 
        session = (WT_SESSION_IMPL *)cbt->iface.session;
 
-       F_CLR(cbt, WT_CBT_RETRY_NEXT | WT_CBT_RETRY_PREV);
-
        if (reenter) {
 #ifdef HAVE_DIAGNOSTIC
                __wt_cursor_key_order_reset(cbt);
```